### ISSUE TITLE
Training time increase between 0.8.0 and 0.9.0 upgrade.

### ISSUE THREAD

**Issue by quant-exchange** [ourownstory/neural_prophet#1629]
We use 2 instances of NP trainings in our pipeline per result and running them back-to-back, historically, we noticed around a 7–8-minute runtime using 0.8.0. However, when upgrading to 0.9.0, we around between 10-12-minute range. Not the biggest deal, but does bring up questions around how future updates could affect training times and explainability around this =).

**Comment by quant-exchange** [comment]
@ourownstory yes, we use a YAML to configure NP with custom learning rate, epochs and batch size.

**Comment by ourownstory** [comment]
Thank you - I just looked at the change set, and the only significant change that I saw was the upgrade to Lightning 2.0. I suspect it may be connected to some of their upstream changes. 

Would you mind profiling the runtimes for your implementation with 0.8.0 and 0.9.0?  That may help us identify potential sources.

**Comment by quant-exchange** [comment]
We will work on setting up a profiling for NP pipeline and report back when we have results.

**Comment by ourownstory** [comment]
Sounds good - FYI we are also planning to add a mini tutorial for profiling: #1628

**Comment by MaiBe-ctrl** [comment]
Opened an issue at Pytorch Lightning for further investigation:  [Link to Issue ](https://github.com/Lightning-AI/pytorch-lightning/issues/20201)

**Comment by quant-exchange** [comment]
I need to add another note here***: I'm using an NVIDIA 4070 RTX laptop GPU for these trainings, but these processes are running 24/7.

So if I do any profiling, wanted to make sure that my requirements are biased towards using that configuration*.

**Comment by MaiBe-ctrl** [comment]
Thank you for sharing! We've conducted profiling on various devices (CPU, GPU, MPS) and observed an average 25% increase in training time after upgrading to Lightning 2.0. It would be great to compare this with your profiling results to see if our findings align.

**Comment by ourownstory** [comment]
> that configuration (GPU acceleration in general)

RE: GPU acceleration: We have also noticed that the training time is higher on GPU than CPU for small to regular size models. Moving data to device takes some extra time. GPU acceleration does however speed up training for larger, i.e. deep NN with many lags/forecasts, and larger batches. Keep this in mind and make sure you do your profiling on a similar setup as you deploy.
(This is independent of Lightning 2.0)

**Comment by ourownstory** [comment]
@quant-exchange As we were able to isolate the issue to be a Pytorch Lighning issue, we are hoping that they may address this eventually. 

Meanwhile, you could experiment with the intersection of different PL versions and Python versions, as there may potentially exist unaffected combinations.

We had rewritten NP to use Lightning about two years ago, in order to replace some of our own or third party implementations with theirs (metrics, learning rate finder) and to make use of some of their other features (tensorboard integration, checkpointing, etc). 
While it would be possible to move away from PL, it would likely mean to strip away a lot of features, or to have a separate implementation, but we currently do not have the bandwidth to support two different versions.

BTW, our latest main version is now memory-optimized, reducing memory use by  up to `n_forecasts`+`n_lags` times.

**Comment by quant-exchange** [comment]
We're finally moving our container to a cloud provider NVIDIA GPU instance! With that said, we'll only do any performance testing/analysis there, which will be a more realistic ENV than GPU laptops!

**Comment by quant-exchange** [comment]
Looking at forecaster.py I see where trainer_config lives. Wanted to see if moving from 32-bit to 16-bit (16-mixed?) for mixed precision could be fruitful for performance:

```code
trainer_config: Optional[dict] = {
    "precision": 16,  # Enable mixed precision for speed-up (16-mixed?)
    "accelerator": "gpu",  # or "cpu" depending on your hardware
    "devices": 1,  # Adjust based on available GPUs
    "profiler": "advanced",  # Use profiler to track performance bottlenecks
},
```

Default for PyTorch Lightning 1.x & 2.x for precision is 32-bit. Is 32-bit the value being used by intention with the NeuralProphet project or are there other values recommended or that have been tested since the 2.x update? 

Is there currently a way to pass the precision value via this syntax?:

```code
model = NeuralProphet(
    n_lags=engine['lags'],
    n_forecasts=engine['ptf'],
    batch_size=engine['batch_size'],
    learning_rate=engine['learning_rate'],
    growth='off',
    yearly_seasonality=False,
    weekly_seasonality=False,
    daily_seasonality=False,
    seasonality_reg=0,
    n_changepoints=0
)
```

**Comment by quant-exchange** [comment]
@MaiBe-ctrl is using >=2.0.0 with the intention of which version: 2.0, 2.1, 2.2, 2.3 or 2.4 (>=2.0.0 depending on the system source could pull any 2.x version potentially making that testing instable)? For example, if we edited this to be >=2.4.0 we could ensure it's of the latest 2.4.x version*.

For me doing a: ```pip show pytorch-lightning```:

Name: pytorch-lightning
Version: 2.4.0
Summary: PyTorch Lightning is the lightweight PyTorch wrapper for ML researchers. Scale your models. Write less boilerplate.
Home-page: https://github.com/Lightning-AI/lightning
Author: Lightning AI et al.
Author-email: pytorch@lightning.ai
License: Apache-2.0
Location: /usr/local/lib/python3.9/site-packages
Requires: fsspec, lightning-utilities, packaging, PyYAML, torch, torchmetrics, tqdm, typing-extensions
Required-by: neuralprophet

I do get 2.4.0 (but just to be safe might be a good idea to make it >=2.4.0 =).

**Comment by quant-exchange** [comment]
```model.fit(training_data, freq=engine['freq'], epochs=engine['epochs'], minimal=True)```: Adding ```minimal=True``` from what I can tell in local testing makes the training process significantly quicker! Going to add this update to our server implementation for testing now*.

**Comment by quant-exchange** [comment]
This update cut the training time by up to 50% on the cloud server!

**Comment by quant-exchange** [comment]
Since this is not an issue for me now, I think I'll close it and let the ticket with PyTorch-Lightning be the performance reference!
